### PR TITLE
Fix resolv.conf during Azure host servicing

### PR DESCRIPTION
**Pull request by s-fairchild**
### Which issue this PR addresses:

Fixes

https://issues.redhat.com/browse/ARO-3474
https://issues.redhat.com/browse/ARO-3468

### What this PR does / why we need it:

If the network interfaces are brought up/down without restarting NetworkManager, our resolv.conf will be overwritten by NetworkManager. This can occur during [Azure Host Servicing](https://learn.microsoft.com/en-us/azure/developer/intro/hosting-apps-on-azure) events.

See [NetworkManager-dispatcher](https://networkmanager.dev/docs/api/latest/NetworkManager-dispatcher.html) documentation for more information, such as available actions.  

The second commit cleans up the dnsmasq controller package by embedding all Go templates, and moving them into their own files within `pkg/operators/controllers/dnsmasq/scripts`.

**Operator flag has been set to true**

### Test plan for issue:

Tested on a dev cluster with, I have created a machineconfig locally containing the changes and applied them to both master and worker nodes successfully.
The restart script works and has been tested with simulated Azure host servicing in this [jira ticket](https://issues.redhat.com/browse/ARO-3440).

### CODE REVIEW

File: vendor/github.com/openshift/installer/pkg/aro/dnsmasq/dnsmasq.go
Comment (tsatam) [inline vendor/github.com/openshift/installer/pkg/aro/dnsmasq/dnsmasq.go:111]: As of #2932, this file has been duplicated directly into the ARO-RP codebase rather than referenced from the vendored dependency.

If I'm understanding this correctly, these changes will need to be made to both:
- ARO-RP: https://github.com/Azure/ARO-RP/blob/master/pkg/operator/controllers/dnsmasq/dnsmasq.go
- installer-aro: https://github.com/jewzaam/installer-aro/blob/release-4.11-azure/pkg/aro/dnsmasq/dnsmasq.go

as this specific vendored file is no longer referenced by our operator reconciler.

Comment (s-fairchild) [reply vendor/github.com/openshift/installer/pkg/aro/dnsmasq/dnsmasq.go]: @tsatam Thank you for bringing this to my attention!

**Comment by s-fairchild** [comment]
> Is it possible to emit a log when dnsmasq gets restarted by systemctl? Ultimately ending up in Kusto for observability

@swiencki Yes, I can add that to the script. I'll verify, but `logger` should be installed and we can emit a system log there.
What log level do we want? and any specific information, we can include [anything available here](https://networkmanager.dev/docs/api/latest/NetworkManager-dispatcher.html) in the logs.

**Comment by swiencki** [comment]
> > Is it possible to emit a log when dnsmasq gets restarted by systemctl? Ultimately ending up in Kusto for observability
> 
> @swiencki Yes, I can add that to the script. I'll verify, but `logger` should be installed and we can emit a system log there. What log level do we want? and any specific information, we can include [anything available here](https://networkmanager.dev/docs/api/latest/NetworkManager-dispatcher.html) in the logs.

Oh I  get to choose!

If we can emit logs from [NetworkManager-Dispatcher](https://networkmanager.dev/docs/api/latest/NetworkManager-dispatcher.html) then;

up - The interface has been activated.
down - The interface has been deactivated.
connectivity-change - The network connectivity state has changed (no connectivity, went online, etc).

Maybe reapply, not sure what that is?

env vars we might be able to craft a print similar to;

$TIMESTAMP - Current state: CONNECTIVITY_STATE: $NM_DISPATCHER_ACTION on $DEVICE_IFACE:DEVICE_IP_IFACE (IP4_NUM_ADDRESSES/IP4_GATEWAY)

The following outputs might also be useful:
IP4_NUM_ROUTES | The variable contains the number of IPv4 routes the script may expect.
IP4_NAMESERVERS | The variable contains a space-separated list of the DNS servers.
IP4_DOMAINS | The variable contains a space-separated list of the search domains.
DHCP4_<dhcp-option-name> | If the connection used DHCP for address configuration, the received DHCP configuration is passed in the environment using standard DHCP option names, prefixed with "DHCP4_", like "DHCP4_HOST_NAME=foobar".

cat of; 
dnsmasq.conf
/etc/resolv.conf
/etc/resolv.conf.dnsmasq

All of the above might be a bit verbose but could provide much needed insight..

**Comment by ventifus** [comment]
@swiencki NM normally logs interface up/down events and also when it gets a new DHCP lease:

```
master-0 NetworkManager[1359]: <info>  [1687362059.7572] device (eth0): carrier: link connected
master-0 NetworkManager[1359]: <info>  [1687362059.7577] manager: (eth0): new Ethernet device (/org/freedesktop/NetworkManager/Devices/3)
master-0 NetworkManager[1359]: <info>  [1687362059.7584] manager: (eth0): assume: will attempt to assume matching connection 'Wired Connection' (a9d5d6a9-ba0d-407c-b174-b5e6216cb8b5) (indicated)
master-0 NetworkManager[1359]: <info>  [1687362059.7584] device (eth0): state change: unmanaged -> unavailable (reason 'connection-assumed', sys-iface-state: 'assume')
master-0 NetworkManager[1359]: <info>  [1687362059.7591] device (eth0): state change: unavailable -> disconnected (reason 'connection-assumed', sys-iface-state: 'assume')
master-0 NetworkManager[1359]: <info>  [1687362059.7599] device (eth0): Activation: starting connection 'Wired Connection' (a9d5d6a9-ba0d-407c-b174-b5e6216cb8b5)
master-0 NetworkManager[1359]: <info>  [1687362059.7646] device (eth0): state change: disconnected -> prepare (reason 'none', sys-iface-state: 'assume')
master-0 NetworkManager[1359]: <info>  [1687362059.7650] device (eth0): state change: prepare -> config (reason 'none', sys-iface-state: 'assume')
master-0 NetworkManager[1359]: <info>  [1687362059.7730] device (eth0): state change: config -> ip-config (reason 'none', sys-iface-state: 'assume')
master-0 NetworkManager[1359]: <info>  [1687362059.7735] dhcp4 (eth0): activation: beginning transaction (timeout in 90 seconds)
master-0 NetworkManager[1359]: <info>  [1687362059.7837] dhcp4 (eth0): state changed new lease, address=10.0.0.7
master-0 NetworkManager[1359]: <info>  [1687362059.7840] policy: set 'Wired Connection' (eth0) as default for IPv4 routing and DNS
master-0 NetworkManager[1359]: <info>  [1687362059.7901] device (eth0): state change: ip-config -> ip-check (reason 'none', sys-iface-state: 'assume')
master-0 NetworkManager[1359]: <info>  [1687362059.8752] device (eth0): state change: ip-check -> secondaries (reason 'none', sys-iface-state: 'assume')
master-0 NetworkManager[1359]: <info>  [1687362059.8754] device (eth0): state change: secondaries -> activated (reason 'none', sys-iface-state: 'assume')
master-0 NetworkManager[1359]: <info>  [1687362059.8760] device (eth0): Activation: successful, device activated.
```

**Comment by swiencki** [comment]
> @swiencki NM normally logs interface up/down events and also when it gets a new DHCP lease:

Beauty, +1 on limiting the amount of duplicate logging

**Comment by s-fairchild** [comment]
@swiencki @ventifus I've pushed some changes to the restart script that improve logging information. Here's some example output

```bash
sh-4.4# ./20-dnsmasq-restart eth0 up
+ interface=eth0
+ action=up
+ '[' eth0 == 'eth*' ']'
+ '[' up == up ']'
+ log 'up happened on eth0, connection state is now '
+ logger -i ./20-dnsmasq-restart -t '99-DNSMASQ-RESTART SCRIPT' 'up happened on eth0, connection state is now '
+ log 'restarting dnsmasq now'
+ logger -i ./20-dnsmasq-restart -t '99-DNSMASQ-RESTART SCRIPT' 'restarting dnsmasq now'
+ systemctl restart dnsmasq
+ log 'dnsmasq successfully restarted'
+ logger -i ./20-dnsmasq-restart -t '99-DNSMASQ-RESTART SCRIPT' 'dnsmasq successfully restarted'
++ cat /etc/resolv.conf
+ log '/etc/resolv.conf output
    # Generated for dnsmasq.service - should point to self
search g323k5hfzoau3c2bwxrpslve2f.bx.internal.cloudapp.net 
nameserver 10.0.18.8'
+ logger -i ./20-dnsmasq-restart -t '99-DNSMASQ-RESTART SCRIPT' '/etc/resolv.conf output
    # Generated for dnsmasq.service - should point to self
search g323k5hfzoau3c2bwxrpslve2f.bx.internal.cloudapp.net 
nameserver 10.0.18.8'
++ cat /etc/resolv.conf.dnsmasq
+ log '/etc/resolv.conf.dnsmasq output
    # Generated for dnsmasq.service - should contain DHCP configured DNS
search g323k5hfzoau3c2bwxrpslve2f.bx.internal.cloudapp.net 
nameserver 168.63.129.16'
+ logger -i ./20-dnsmasq-restart -t '99-DNSMASQ-RESTART SCRIPT' '/etc/resolv.conf.dnsmasq output
    # Generated for dnsmasq.service - should contain DHCP configured DNS
search g323k5hfzoau3c2bwxrpslve2f.bx.internal.cloudapp.net 
nameserver 168.63.129.16'
+ exit 0
```
An example of an action not causing a restart:
```bash
sh-4.4# ip link set eth0 down; ip link set eth0 up
Jun 28 20:39:51 steven-cluster-bh8mr-master-0 nm-dispatcher[219385]: + interface=eth0
Jun 28 20:39:51 steven-cluster-bh8mr-master-0 nm-dispatcher[219385]: + action=dhcp4-change
Jun 28 20:39:51 steven-cluster-bh8mr-master-0 nm-dispatcher[219385]: + '[' eth0 == 'eth*' ']'
Jun 28 20:39:51 steven-cluster-bh8mr-master-0 nm-dispatcher[219385]: + '[' dhcp4-change == up ']'
Jun 28 20:39:51 steven-cluster-bh8mr-master-0 nm-dispatcher[219385]: + '[' dhcp4-change == down ']'
Jun 28 20:39:51 steven-cluster-bh8mr-master-0 nm-dispatcher[219385]: + exit 0
```

**Comment by yjst2012** [comment]
LGTM, but I would like another codeowner to approve, re-run e2e to see if it is a flake, thanks for putting this up!

**Review by petrkotas — APPROVED** [review]
Nice! I like the explanation which made the review easier. LGTM

**Comment by UlrichSchlueter** [comment]
LGTM but would like to hold this one until we are clear on the rollout impact (PUCM run, worker reboots)

**Comment by s-fairchild** [comment]
Thanks for the reviews everyone!
Latest push has small changes, I  realized I hadn't set the file mode, and improved the if conditional matching of eth* (had all ||, was missing an &&).

I've tested by running `go run -tags aro ./cmd/aro operator master` against my cluster and all works as expected, machineconfigs render and the script executes.

**Review by s-amann — APPROVED** [review]

**Review by cadenmarchese — APPROVED** [review]

File: pkg/operator/controllers/dnsmasq/dnsmasq.go
Comment (ventifus) [inline pkg/operator/controllers/dnsmasq/dnsmasq.go:127]: In the test, the interface that gets unplugged / replugged (the VF interface) is called enP53531s1. Maybe we look for interfaces called `e*` or specifically eth* plus enP*?

File: pkg/operator/controllers/dnsmasq/dnsmasq.go
Comment (ventifus) [inline pkg/operator/controllers/dnsmasq/dnsmasq.go:130]: `systemctl restart` will start dnsmasq if it's not running. Could we end up racing eth0's DHCP on initial boot potentially ending up with nothing in /etc/resolv.conf.dnsmasq? Let's `systemctl try-restart` here instead so that it will only restart if it's already running.

**Review by s-amann — APPROVED** [review]

**Comment by cadenmarchese** [comment]
@s-fairchild mind TAL at the e2e failure? It looks related to your change: https://github.com/Azure/ARO-RP/pull/2995/checks?check_run_id=15211124105

**Comment by s-fairchild** [comment]
@cadenmarchese Yes i'll look into them, may have to run e2e locally on Monday to see what's going on.

File: pkg/api/defaults.go
Comment (bennerv) [inline pkg/api/defaults.go:72]: I think this will break all the existing clusters that use egress lockdown.  

I was hoping more for an approach where there's a new feature flag and a new file that's laid down, either within the dnsmasq controller, or a separate controller.

File: pkg/operator/controllers/dnsmasq/dnsmasq.go
Comment (bennerv) [inline pkg/operator/controllers/dnsmasq/dnsmasq.go:113]: move this to a new file or string which is applied ad-hoc.

File: pkg/operator/controllers/dnsmasq/dnsmasq.go
Comment (bennerv) [inline pkg/operator/controllers/dnsmasq/dnsmasq.go:233]: before this, drop in a feature flag above this on whether or not we enable this.

**Review by bennerv — DISMISSED** [review]
I was thinking of a new feature flag.  We shouldn't disable dnsmasq feature for existing clusters, hence the new feature flag to put these behind.  

Then we can apply it ad-hoc to clusters.

**Comment by s-fairchild** [comment]
@bennerv I misunderstood, thanks for explaining. I've updated the branch with a new operator flag to enable/disable these changes specifically.
I want to add an e2e test for the new flag soon so it actually gets tested.

**Review by kimorris27 — COMMENTED** [review]
Looking good! I made two small recommendations on the restart script.

I also have an open-ended question. I saw in the description of https://issues.redhat.com/browse/ARO-3474 the concern of ensuring that customer changes to the cluster that write to `resolv.conf` are still allowed. The solution implemented in this PR will preserve ARO-specific `resolv.conf` configurations, but do you think it's possible to preserve customer-initiated changes as well?

I'd imagine that when Azure host servicing happens, we could still end up getting DNS-related incidents because of the loss of that configuration. These incidents would be less severe than the one that got this work started though, since the critical stuff from an SRE perspective is the ARO-specific configuration.

**Comment by s-fairchild** [comment]
> Looking good! I made two small recommendations on the restart script.
> 
> I also have an open-ended question. I saw in the description of https://issues.redhat.com/browse/ARO-3474 the concern of ensuring that customer changes to the cluster that write to `resolv.conf` are still allowed. The solution implemented in this PR will preserve ARO-specific `resolv.conf` configurations, but do you think it's possible to preserve customer-initiated changes as well?
> 
> I'd imagine that when Azure host servicing happens, we could still end up getting DNS-related incidents because of the loss of that configuration. These incidents would be less severe than the one that got this work started though, since the critical stuff from an SRE perspective is the ARO-specific configuration.

@ventifus Correct me if I'm wrong here please..
It depends on how customers are making these custom DNS changes. If they are following [this documentation](https://learn.microsoft.com/en-us/azure/openshift/howto-custom-dns) I think this change won't interfere with custom DNS.
Customer's should also be able to set their own operator flags I believe(?), so they could disable this restart script from being added to the dnsmasq machine configs. With the dnsmasq restart flag enabled, this should happen.
If a customer tries to manually set DNS on the nodes via editing `/etc/resolv.conf`, then this will overwrite those changes. If the customer sets custom DNS and DOMAIN settings via `nmcli`, then the next time `dnsmasq` gets restarted those settings will be used. For a short time anyways. Once the DHCP lease get's renewed the DNS settings made with `nmcli` will get overwritten, followed by our Pre scripts getting the new addresses.

To summarize, if the customer sets custom DNS via Azure's DHCP server settings, this change ensures that that address is always used.

An off topic question, would anyone mind if I moved the dnsmasq operator's templates to actual files and then used embed? I'm going to do this in a second commit and I can delete if we don't want that.

**Comment by kimorris27** [comment]
> To summarize, if the customer sets custom DNS via Azure's DHCP server settings, this change ensures that that address is always used.

Cool, this explanation makes sense to me.

> An off topic question, would anyone mind if I moved the dnsmasq operator's templates to actual files and then used embed? I'm going to do this in a second commit and I can delete if we don't want that.

👍🏻 from me on this change.

**Comment by ventifus** [comment]
@kimorris27 If a customer tries to edit resolv.conf via MachineConfig, that won't work well, since NetworkManager rewrites that file on boot, and aro-dnsmasq-pre.sh swaps it with resolv.conf.dnsmasq on service start. I'm not sure if there's any other avenue of modifying that file that wouldn't have conflicts.

**Comment by kimorris27** [comment]
>@kimorris27 If a customer tries to edit resolv.conf via MachineConfig, that won't work well, since NetworkManager rewrites that file on boot, and aro-dnsmasq-pre.sh swaps it with resolv.conf.dnsmasq on service start. I'm not sure if there's any other avenue of modifying that file that wouldn't have conflicts.

Ah, I see. Since modifying the MachineConfigs is unsupported, I don't think I'm worried about that case. As long as the customer's configuration is maintained when they modify it in the supported way, that seems good enough.

File: pkg/operator/controllers/dnsmasq/scripts/99-dnsmasq-restart.gotmpl
Comment (cadenmarchese) [inline pkg/operator/controllers/dnsmasq/scripts/99-dnsmasq-restart.gotmpl:30]: It looks like we are proceeding from here no matter what (if the NIC is up, we proceed, if the NIC is down, we proceed), so do we need this `if` statement?

File: pkg/operator/controllers/dnsmasq/scripts/99-dnsmasq-restart.gotmpl
Comment (cadenmarchese) [inline pkg/operator/controllers/dnsmasq/scripts/99-dnsmasq-restart.gotmpl:33]:
```suggestion
    log_dns_files
```
Nit, since it doesn't looks like we are doing a comparison.

File: pkg/operator/controllers/dnsmasq/scripts/99-dnsmasq-restart.gotmpl
Comment (cadenmarchese) [inline pkg/operator/controllers/dnsmasq/scripts/99-dnsmasq-restart.gotmpl:16]:
```suggestion
log_dns_files() {
```

File: pkg/operator/controllers/dnsmasq/scripts/99-dnsmasq-restart.gotmpl
Comment (cadenmarchese) [inline pkg/operator/controllers/dnsmasq/scripts/99-dnsmasq-restart.gotmpl:35]: Where do we compare the current dnsmasq configuration with the "correct" one? Or, are we just restarting whenever we get a machine config change from the operator?

File: pkg/api/defaults.go
Comment (cadenmarchese) [inline pkg/api/defaults.go:73]: Since our next PUCM will already cause reboots, we can merge with this enabled:
```suggestion
		"aro.restartdnsmasq.enabled":               flagTrue,
```

File: pkg/operator/controllers/dnsmasq/scripts/99-dnsmasq-restart.gotmpl
Comment (s-fairchild) [reply pkg/operator/controllers/dnsmasq/scripts/99-dnsmasq-restart.gotmpl:30]: There are a number of other things it can be, checkout the documentation for available actions: https://networkmanager.dev/docs/api/latest/NetworkManager-dispatcher.html

File: pkg/operator/controllers/dnsmasq/scripts/99-dnsmasq-restart.gotmpl
Comment (s-fairchild) [reply pkg/operator/controllers/dnsmasq/scripts/99-dnsmasq-restart.gotmpl:35]: We are restarting regardless of the differences, and logging them.

**Comment by kimorris27** [comment]
This LGTM! I'll revisit after E2E is passing.

File: pkg/operator/controllers/dnsmasq/dnsmasq.go
Comment (bennerv) [inline pkg/operator/controllers/dnsmasq/dnsmasq.go:28]:
```suggestion
	t := template.Must(template.New(configFileName).Parse(configFile))

**Review by bennerv — CHANGES_REQUESTED** [review]

File: pkg/operator/controllers/dnsmasq/machineconfig_controller.go
Comment (kimorris27) [inline pkg/operator/controllers/dnsmasq/machineconfig_controller.go:58]: nit: Can we assign the operator flag's value to the intermediate variable `restartDnsmasq` and use it both here and in the call to `reconcileMachineConfigs`? Just so the three controllers that reference the flag match each other in this way.

**Review by kimorris27 — COMMENTED** [review]
Looking good! One nit.

**Review by cadenmarchese — APPROVED** [review]
Steps taken to test this, in addition to the CI checks here:

- Shared cluster
  - Confirmed that we could run this new operator in an existing cluster without any restarts (until a PUCM to update the operator feature flags goes out)
- Local development
  - Built a new cluster with these changes to ensure that creation was successful and the cluster custom resource got the new feature flag
  - [Ran the operator out of cluster](https://github.com/Azure/ARO-RP/blob/master/docs/operators.md#how-to-run-the-operator-locally-out-of-cluster) with these changes to confirm that the dispatcher.d script was created on all nodes (`oc get node -o name | xargs -I {} oc debug {} -- chroot /host sh -c 'cat /etc/NetworkManager/dispatcher.d/99-dnsmasq-restart'`)
- INT
  - Confirmed that the dispatcher.d script was created on all nodes in a new cluster created in INT

File: pkg/operator/controllers/dnsmasq/machineconfig_controller.go
Comment (s-fairchild) [reply pkg/operator/controllers/dnsmasq/machineconfig_controller.go:58]: Yes, I hadn't really thought about it that way. I choice not to use a variable here simply because it wasn't necessary. But I like your suggestion of continuity, I'll make a follow up PR.

**Pull request merged**
